### ISSUE TITLE
Add a "Build" section to the README

### ISSUE THREAD

**Issue by mateor** [twitter/finagle#511]
I cloned Finagle to do some experimenting. I was surprised to see that my clone did not compile when  following the directions from [Scala School](https://twitter.github.io/scala_school/finagle.html). 

      $ cd finagle
      $ ./sbt "project finagle-http" console
       ...build output...
      scala>

The above snippet failed for the same reason that the `develop` branch of Travis CI is failing, there are no published `SNAPSHOT` versions of the `com.twitter` dependencies.

I eventually found the supported workflow for contributors under CONTRIBUTING.md and was able to move forward.  But I suspect that some Finagle-curious people might have just given up.

Cloning a codebase from github checks out the branch that the repo has set as 'Default'.  For Finagle, the default branch is the 'develop' branch.  I opened this issue to suggest that [the default branch of 
this project be changed to 'master'](https://help.github.com/articles/setting-the-default-branch/) since it  is actually expected to compile without needing to clone + compile the dependency graph.

Thanks!

**Comment by mosesn** [comment]
@mateor we want `develop` to be the default branch because it has the latest code, so I'm not sure we want to switch the default to be `master`.  Do you think we could improve the documentation to make this more clear?  In general, our assumption is that people who want to build the project want to do it because they are contributing, so they want the latest version.

**Comment by mateor** [comment]
I actually figured it out pretty quickly. The CONTRIBUTING doc seems up to date, although I didn't follow it - I just switched to master. It would be nice to have the Scala School docs work but of course that isn't your project.

I opened the issue on the chance that you didn't know that a git clone wouldn't compile by default. If it is a considered choice, I think #wontfix is probably okay.

My feedback if you keep the 'develop' branch would be to add a 'build' or 'run' section to the README that links to the steps needed to get up and running. I don't usually expect to look in CONTRIBUTING for build instructions, that usually details the review process in my experience.

I appreciate the reply and that you are doing this development in public in the first place, so thanks!

**Comment by mateor** [comment]
Although I will add, I don't know if I agree with the bias towards potential contributors. Anyone who is going to contribute will have to do a good bit of work/configuration that dominates the time it takes to check out a development branch. I don't know if that convenience is worth the cost of having most people's first experience with Finagle being an error message about missing dependencies.

**Comment by mosesn** [comment]
OK, let's mark the remediation item here as "Add a Build section to the README" and it should encourage people to build off of master.  I'm going to mark this as a starter ticket for now.  Thanks for filing a ticket!

**Comment by mosesn** [comment]
@mateor that's fair.  we would like to have a tool which makes it easier to set up the various dependent repositories, maybe a dbuild setup, but we haven't had the bandwidth to set it up.  maybe if we actually get that set up, we'll be able to switch the default to `master`.  the other alternative we've been looking at is publishing SNAPSHOT builds to maven central so that the `develop` branch builds, but we also haven't had the opportunity to do that yet.

**Comment by adriancole** [comment]
Ran into this myself. I think the problem is that the build instructions aren't rooted here. For example, one shouldn't need to go to scala school or scrooge to figure out how to build finagle. On that note, the instructions in scrooge don't work either due to a cyclic dependency.

Ex. even if you skip tests like below, there's eventually a failure publishing scrooge locally due to a dependency on finagle. And you can't build finagle due to a missing dependency on scrooge.

```bash
# after publishLocal on util, from scrooge, this will eventually fail.
./sbt "set test in publishLocal := {}" publishLocal
--snip--
[error] (scrooge-generator-tests/*:update) sbt.ResolveException: unresolved dependency: com.twitter#finagle-thrift_2.11;6.35.0-SNAPSHOT: not found
```

I'll try building from master as building from develop seems impractical.

**Comment by mosesn** [comment]
@adriancole for what it's worth, scrooge-generator-tests are only tests, and nothing else has a dependency on finagle, but we're aware it's a pain point.  I think this would be better if we had a tool which handled this for you.

**Comment by adriancole** [comment]
>
> @adriancole <https://github.com/adriancole> for what it's worth,
> scrooge-generator-tests are only tests, and nothing else has a dependency
> on finagle, but we're aware it's a pain point. I think this would be better
> if we had a tool which handled this for you.
>
do you know of a syntax to get publishLocal on scrooge to work without it,
then? my sbt-fu might be lacking

**Comment by mosesn** [comment]
It looks like we made a command for this particular purpose: `./sbt +scrooge-publish-local/publishLocal` should work, I think.

**Comment by adriancole** [comment]
getting closer.. it publishes things, except not scrooge-sbt-plugin, which
is what finagle's build depends on

**Comment by mosesn** [comment]
Looks like we treat it specially because it can only be published against 2.10 (since it's a plugin).

`./sbt ++2.10.6 scrooge-generator/publishLocal scrooge-sbt-plugin/publishLocal`

**Comment by olix0r** [comment]
linkerd has a build script for CI'ing against the develop branch that may be helpful: https://github.com/BuoyantIO/linkerd/blob/master/ci/twitter-develop.sh

**Comment by kevinoliver** [comment]
I think we've made this quite a bit better since this issue was opened. The contributing section of README.md points you to CONTRIBUTING.md which in turn has the instructions for building the dependencies. That part is now a 1-liner: `curl -s https://raw.githubusercontent.com/twitter/dodo/develop/bin/build | bash -s -- --no-test finagle`.

Closing this as I believe we've addressed the issues. Thanks for the feedback.
